text/0819-public-router-service.md

```diff
@@ -0,0 +1,242 @@
+---
+Stage: Accepted
+Start Date: 2022-05-12
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/819
+---
+
+<!--- 
+Directions for above: 
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Public `RouterService` import
+
+## Summary
+
+This RFC proposes adding a public import for [`RouterService`](https://api.emberjs.com/ember/release/classes/RouterService)
+making it extensible by addons, removing the external link assertion from `<LinkTo />` and rebuilding the `<LinkTo />` component
+to rely entirely on the public routing service instead of the internal `-router`.
+
+## Motivation
+
+Today, users of [`ember-engines`](https://github.com/ember-engines/ember-engines) are required to utilize separate
+custom components, helpers and services for linking. This leaves the experience feeling sub-par, in addition to introducing
+extra maintenance friction attempting to keep these primitives aligned with those in `ember-source`.
+
+By providing a public import for `RouterService`, `ember-engines` would be able to provide its own router service
+implementation extending from the base implementation. Combined with minor changes to the engines-specific behavior of
+`link-to`, this change would allow us to fill the existing DX gap when it comes to navigating between an Engine
+and a host App *and* allow engines to eliminate use of the legacy components required to support 4.x versions of Ember.
+
+These proposed changes would further allow `ember-engines` to deprecate the `<LinkToExternal />`
+component provided by `ember-engines` package. As a result, there will be less
+engine specific concepts and code would be more portable between regular applications and engines.
+
+Once this RFC gets implemented, following will be possible in user defined Ember Engines:
+
+* links to external routes in templates within the engine rendered via `<LinkTo />` component:
+
+  ```hbs
+  <LinkTo @route="external-route">Go Somewhere</LinkTo>
+  ```
+
+* using `RouterService` methods for redirects to the external routes:
+
+  ```javascript
+  # addon/components/my-component.js
+  export default class extends Component {
+    @service router;
+    @action clickLink () {
+      this.router.transitionTo('external-route');
+    }
+  }
+  ```
+
+**Bonus: improved DX for addons**
+
+This RFC seeks to provide a solution that also solves another DX issue faced by engines developers.
+
+Authors of large scale applications often create an addon(s) to contain reusable components that can
+be rendered either in the context of the host App or Engine. In such scenarios an extra guard needs
```

Comment:
this is also solved by using hrefs instead of route/model/query args of the default link-to.
with an `a` tag with `preventDefault`d click handler, we can use the router service to detect when an href is in the route, or in an external app, and appropriately "do the right thing.

I like this approach because it means we:
 - don't need extra components for each type of usage
 - uses the existing href knowledge that everyone already understands with `<a>`
 - aligns with:  (the spirit of): https://github.com/emberjs/ember.js/pull/19271 and https://github.com/emberjs/rfcs/pull/391
 
 here is a demo: https://codesandbox.io/s/custom-link-component-dgbxl?file=/app/components/link.hbs
 (this demo is incomplete, but gives the gist).
 (except codesandbox doesn't work so... :shrug: (this sandbox has been untouched since 3.21, and the dependency checker is preventing start up... idk what's going on. ))
 
 where improvements could be made here where `<LinkTo>` is a bit better at:
  - query-param only transitions
  - sticky query params

Reply:
though, unifying on LinkTo is still a good improvement!

Reply:
This RFC is what would make those router helpers and the href approach work correctly for engines as well.